File: templates/problems/11_graphs/12_redundant_connection/solution.go
package redundant_connection

// solveFindRedundantConnection finds the edge that creates a cycle using Union-Find.
// Time: O(n * alpha(n)) ~= O(n), Space: O(n)
func solveFindRedundantConnection(edges [][]int) []int {
	n := len(edges)
	parent := make([]int, n+1)
	rank := make([]int, n+1)
	for i := range parent {
		parent[i] = i
	}

	var find func(x int) int
	find = func(x int) int {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}

	for _, e := range edges {
		px, py := find(e[0]), find(e[1])
		if px == py {
			return e
		}
		if rank[px] < rank[py] {
			px, py = py, px
		}
		parent[py] = px
		if rank[px] == rank[py] {
			rank[px]++
		}
	}
	return nil
}

File: templates/problems/11_graphs/12_redundant_connection/python/solution.py
class UnionFind:
    """Union-Find (Disjoint Set) with path compression and union by rank."""

    def __init__(self, n: int) -> None:
        self.parent = list(range(n + 1))
        self.rank = [0] * (n + 1)

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]


def solve_find_redundant_connection(edges: list[list[int]]) -> list[int]:
    """Find the edge that creates a cycle using Union-Find.

    Time: O(n * alpha(n)) ~= O(n), Space: O(n).
    """
    n = len(edges)
    uf = UnionFind(n)

    for e in edges:
        px, py = uf.find(e[0]), uf.find(e[1])
        if px == py:
            return e
        if uf.rank[px] < uf.rank[py]:
            px, py = py, px
        uf.parent[py] = px
        if uf.rank[px] == uf.rank[py]:
            uf.rank[px] += 1
    return []
